src/core/memory.d

```diff
@@ -810,6 +810,106 @@ struct GC
     }
 }
 
+/**
+ * Pure variants of C's memory allocation functions `malloc`, `calloc`, and
+ * `realloc` that achieve purity by aborting the program on failure so
+ * they never visibly change errno.
+ *
+ * The functions may terminate the program using `onOutOfMemoryError` or
+ * `assert(0)`. These functions' purity guarantees no longer hold if
+ * the program continues execution after catching AssertError or
+ * OutOfMemoryError.
+ *
+ * A similar fail-fast variant of C's memory deallocation function `free`
+ * is not provided due to the lack of a portable and efficient way to test
+ * for failure. If purity is required `pureFree` may be used.
+ *
+ * See_Also:
+ *     $(LREF pureMalloc), $(LREF pureCalloc), $(LREF pureRealloc),
+ *     and $(LREF pureFree) which achieve purity by saving and restoring
+ *     errno.
+ *     $(LINK2 https://dlang.org/library/object/error.html, Error class documentation)
+ *     on the danger of continued execution after catching an Error.
+ */
+void* assertMalloc()(size_t size) @nogc nothrow pure @trusted
+{
+    void* ret = fakePureMalloc(size);
+    if (!ret && size)
+    {
+        version (D_Exceptions)
+        {
+            import core.exception : onOutOfMemoryError;
+            onOutOfMemoryError();
+        }
+        else assert(0, "Memory allocation failed");
+    }
+    return ret;
+}
+/// ditto
+void* assertCalloc()(size_t nmemb, size_t size) @nogc nothrow pure @trusted
+{
+    void* ret = fakePureCalloc(nmemb, size);
+    if (!ret && nmemb && size)
+    {
+        version (D_Exceptions)
+        {
+            import core.exception : onOutOfMemoryError;
+            onOutOfMemoryError();
+        }
+        else assert(0, "Memory allocation failed");
+    }
+    return ret;
+}
+/// ditto
+void* assertRealloc()(void* ptr, size_t size) @nogc nothrow pure @system
```

Comment:
Since realloc() free's memory, it cannot ever be considered pure.

Reply:
> Since realloc() free's memory, it cannot ever be considered pure.

We have a [`pureFree`](https://dlang.org/phobos/core_memory.html#.pureFree), and [`GC.free`](https://dlang.org/phobos/core_memory.html#.GC.free) is also marked as `pure`.

`pure` functions are allowed to modify their parameters. Freeing is never `@safe`, but in what way is it less pure than other forms of mutation?

Reply:
There is also already a [`pureRealloc`](https://dlang.org/phobos/core_memory.html#.pureRealloc).

https://github.com/dlang/druntime/blob/db2910ee7425f91dc6fe38f149f4b8eaa3b88f7f/src/core/memory.d#L840-L847